tla-bmcmt/src/test/scala/at/forsyte/apalache/tla/bmcmt/CrossTestEncodings.scala

```diff
@@ -0,0 +1,245 @@
+ *
+ * {{{
+ * ------ MODULE Verifier --------
+ * VARIABLES
+ *   \* @type: ${witnessType};
+ *   result,
+ *   \* @type: Bool;
+ *   found
+ *
+ * Init ==
+ *   /\ result \in ${witnesses}
+ *   /\ result = ${witness}
+ *   /\ found \in BOOLEAN
+ *
+ * Next ==
+ *   UNCHANGED <<result, found>>
+ *
+ * NotFound ==
+ *   ~found
+ * ======================
+ * }}}
+ *
+ * When checking the invariant `NotFound` of the spec `Verifier` ''with a different encoding'', the "verifier encoding",
+ * we can have two possible outcomes:
+ *
+ *   - There is a counterexample. We compare the value of `result` to witness. They should match. If they don't, there
+ *     is a bug in one of the encodings.
+ *   - There is no counterexample. There is a bug in one of the encodings.
+ */
+
+trait CrossTestEncodings extends AnyFunSuite {
+  // Override [[AnyFunSuite.withFixture]] to set the encodings.
+  protected var oracleEncoding: SMTEncoding = _
+  protected var verifierEncoding: SMTEncoding = _
+
+  /**
+   * Rewrite elem \in Set ~~> \E elem$selected \in Set: elem' := elem$selected.
+   */
+  private def rewriteElemInSet(elem: NameEx, set: TlaEx): TlaEx = {
+    val types = Map("e" -> elem.typeTag.asTlaType1(), "b" -> BoolT1())
+    val selected = tla.name(elem.name + "$selected").typed(types, "e")
+    tla
+      .apalacheSkolem(tla.exists(selected, set, tla.assign(tla.prime(elem) ? "e", selected) ? "b") ? "b")
+      .typed(types, "b")
+  }
+
+  /**
+   * Facade for running bounded model-checking through [[SeqModelChecker]].
+   *
+   * @param moduleName
+   *   name of the TLA+ module to check
+   * @param variableDecls
+   *   TLA+ variable declarations
+   * @param initTrans
+   *   init transition, see examples in [[getWitness]] and [[verify]]
+   * @param nextTrans
+   *   next transition, see examples in [[getWitness]] and [[verify]]
+   * @param invExpr
+   *   invariant to check
+   * @param smtEncoding
+   *   SMT encoding to use inside the model checker
+   * @return
+   *   a counterexample returned as VarNames -> Expressions binding in the initial state, i.e., after executing
+   *   [[initTrans]].
+   */
+  private def modelCheckAndGetCex(
+      moduleName: String,
+      variableDecls: Seq[TlaDecl],
+      initTrans: List[TlaEx],
+      nextTrans: List[TlaEx],
+      invExpr: TlaEx,
+      smtEncoding: SMTEncoding): Map[String, TlaEx] = {
+    // prepare the invariant
+    val invDecl = tla.declOp("Inv", invExpr).as(OperT1(Nil, BoolT1()))
+    val decls = variableDecls :+ invDecl
+    val invariantsAndNegations = List((invExpr, tla.not(invExpr).as(BoolT1())))
+    val verificationConditions = CheckerInputVC(invariantsAndNegations)
+
+    val module = TlaModule(moduleName, decls)
+
+    val checkerInput = new CheckerInput(module, initTrans, nextTrans, None, verificationConditions)
+    val checkerParams = new ModelCheckerParams(checkerInput, 0)
+
+    val solverContext = RecordingSolverContext.createZ3(None, SolverConfig.default)
```

Comment:
```suggestion
    val solverContext = new Z3SolverContext(SolverConfig.default)
```

I don't think you need the recording context.